### DISCUSSION TITLE
Accessing body/payload within HttpRequest

### DISCUSSION THREAD

**Discussion by GetSource1234** [actix/actix-web#2235] [Q&A]
Hi!

I have a question. I am trying to implement middleware that do some checks within HttpRequest. Currently, I see only one option is to get payload within ServiceRequest:

`pub fn into_parts(self) -> (HttpRequest, Payload)`

But it is not really handy, since I need to do some check within ServiceResponse. I am wondering is there any approche to get HttpRequest body/payload within ServiceResponse. Also, wondering why such methods are not implemented in HttpRequest, something like:

```
impl HttpRequest {

    ......
    pub fn payload(&self) -> &Payload {
       &self.0.payload
    }
    
    .....
}
```

P.S.

I see pretty match the same [issue](https://github.com/actix/actix-web/issues/1457), But it is closed, without any handy solution.

**Comment by robjtede** [comment]
This will be easier in v4, for now into_parts is the only way to do this. HttpRequest is designed to not allow payload access.

**Reply by moy2010** [reply]
Do you have any hints for this, @robjtede ? I'm trying to access the payload from the JsonConfig's error handler, but all I get is a reference to the HttpRequest and I can't find any method to get the payload from it :/

**Reply by moy2010** [reply]
P.D., I'm using Actix-web version 4.2.1

**Reply by moy2010** [reply]
Update:

I tried using the `from_request` trait, but I think that using `Payload::None` is not what it wants :smile: :

```
block_on(async {
  println!("The payload was {:#?}", web::Bytes::from_request(req, &mut dev::Payload::None).await.unwrap());
});
```

**Reply by robjtede** [reply]
You can not get access to the payload with only a `&HttpRequest`. This is by design.

**Reply by moy2010** [reply]
Got it. I was afraid of that being the case when I looked into the signature for the `into_payload` method :(
